/*
 * Copyright (c) Huawei Technologies Co., Ltd. 2025. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

syntax = "proto3";

package data_service;

option go_package = "faas-pattern/pkg/common/grpc/pb/data;data";

message PutRequest {
  string          objectId        = 1;
  bytes           objectData      = 2;
  int32           writeMode       = 3;	//	put param, default = 0
  int32           consistencyType = 4;	//	put param, default = 0
  repeated string nestedObjectIds = 5;
  int32           cacheType       = 6;
}

message PutResponse {
  int32           code            = 1;
  string          message         = 2;
}

message GetRequest {
  repeated string objectIds       = 1;
  int64           timeoutMs       = 2;
}

message GetResponse {
  int32           code            = 1;
  string          message         = 2;
  repeated bytes  buffers         = 3;
}

message IncreaseRefRequest {
  repeated string objectIds       = 1;
  string          remoteClientId  = 2;
}

message IncreaseRefResponse {
  int32           code            = 1;
  string          message         = 2;
  repeated string failedObjectIds = 3;
}

message DecreaseRefRequest {
  repeated string objectIds       = 1;
  string          remoteClientId  = 2;
}

message DecreaseRefResponse {
  int32           code            = 1;
  string          message         = 2;
  repeated string failedObjectIds = 3;
}

message KvSetRequest {
  string          key             = 1;
  bytes           value           = 2;
  int32           existence       = 3;	//	set param, default = 0
  int32           writeMode       = 4;	//	set param, default = 0
  uint32          ttlSecond       = 5;	//	set param, default = 0
  int32           cacheType       = 6;
}

message KvSetResponse {
  int32           code            = 1;
  string          message         = 2;
}

message KvMSetTxRequest {
  repeated string keys            = 1;
  repeated bytes  values          = 2;
  int32           existence       = 3;
  int32           writeMode       = 4;
  uint32          ttlSecond       = 5;
  int32           cacheType       = 6;
}

message KvMSetTxResponse {
  int32           code            = 1;
  string          message         = 2;
}

message KvGetRequest {
  repeated string keys            = 1;
  uint32          timeoutMs       = 2;	// default = 0
}

message KvGetResponse {
  int32           code            = 1;
  string          message         = 2;
  repeated bytes  values          = 3;
}

message KvDelRequest {
  repeated string keys            = 1;
}

message KvDelResponse {
  int32           code            = 1;
  string          message         = 2;
  repeated string failedKeys      = 3;
}
